mobilecoind-json/src/data_types.rs

```
@@ -103,6 +108,80 @@ impl From<&mc_mobilecoind_api::GetBalanceResponse> for JsonBalanceResponse {
     }
 }
 
+#[derive(Serialize, Deserialize, Default)]
+pub struct JsonUnspentTxOut {
+    pub tx_out: JsonTxOut,
+    pub subaddress_index: u64,
+    pub key_image: String,
+    pub value: String, // Needs to be String since Javascript ints are not 64 bit.
+    pub attempted_spend_height: u64,
+    pub attempted_spend_tombstone: u64,
+    pub monitor_id: String,
+}
+
+impl From<&mc_mobilecoind_api::UnspentTxOut> for JsonUnspentTxOut {
+    fn from(src: &mc_mobilecoind_api::UnspentTxOut) -> Self {
+        Self {
+            tx_out: src.get_tx_out().into(),
+            subaddress_index: src.get_subaddress_index(),
+            key_image: hex::encode(&src.get_key_image().get_data()),
+            value: src.value.to_string(),
+            attempted_spend_height: src.get_attempted_spend_height(),
+            attempted_spend_tombstone: src.get_attempted_spend_tombstone(),
+            monitor_id: hex::encode(&src.get_monitor_id()),
+        }
+    }
+}
+
+// Helper conversion between json and protobuf
+impl TryFrom<&JsonUnspentTxOut> for mc_mobilecoind_api::UnspentTxOut {
+    type Error = String;
+
+    fn try_from(src: &JsonUnspentTxOut) -> Result<mc_mobilecoind_api::UnspentTxOut, String> {
+        let mut key_image = KeyImage::new();
+        key_image.set_data(
+            hex::decode(&src.key_image)
+                .map_err(|err| format!("Failed to decode key image hex: {}", err))?,
+        );
+
+        // Reconstruct the public address as a protobuf
+        let mut utxo = mc_mobilecoind_api::UnspentTxOut::new();
+        utxo.set_tx_out(mc_api::external::TxOut::try_from(&src.tx_out).unwrap());
```

unwrap is unacceptable here. This should return an error.